Content.Server/Supermatter/EntitySystems/SupermatterSystem.cs

```diff
@@ -0,0 +1,425 @@
+using System;
+using Robust.Shared.Audio;
+using Content.Server.Radiation;
+using JetBrains.Annotations;
+using Robust.Shared.GameObjects;
+using Robust.Shared.IoC;
+using Content.Server.Supermatter.Components;
+using Content.Shared.Body.Components;
+using Content.Server.Ghost.Components;
+using Robust.Shared.Containers;
+using Robust.Shared.Map;
+using Robust.Shared.Maths;
+using Content.Shared.SubFloor;
+using Content.Shared.Damage;
+using Content.Shared.Item;
+using Content.Shared.Tag;
+using Content.Server.Projectiles.Components;
+using Content.Server.Explosion.EntitySystems;
+using Content.Server.Chat.Managers;
+using Content.Shared.Damage.Prototypes;
+using Robust.Shared.Prototypes;
+using Content.Server.Atmos.EntitySystems;
+using Robust.Shared.Player;
+using Content.Shared.Atmos;
+using Robust.Shared.Localization;
+using Robust.Shared.Physics.Dynamics;
+using Content.Shared.FixedPoint;
+
+namespace Content.Server.Supermatter.EntitySystems
+{
+    [UsedImplicitly]
+    public class SupermatterSystem : EntitySystem
+    {
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+        [Dependency] private readonly DamageableSystem _damageable = default!;
+        [Dependency] private readonly AtmosphereSystem _atmosphereSystem = default!;
+        [Dependency] private readonly RadiationSystem _radiationSystem = default!;
+        [Dependency] private readonly IChatManager _chatManager = default!;
+        [Dependency] private readonly ExplosionSystem _explosions = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            SubscribeLocalEvent<SupermatterComponent, StartCollideEvent>(OnCollideEvent);
+        }
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            foreach (var (Supermatter, damageable, Xform) in EntityManager.EntityQuery<SupermatterComponent, DamageableComponent, TransformComponent>())
+            {
+                HandleBehavior(Supermatter.Owner, frameTime, Supermatter, damageable, Xform);
+            }
+        }
+
+        /// <summary>
+        /// Handle outputting radiation based off enery, damage, and gas mix
+        /// </summary>
+        public void HandleRads(EntityUid Uid, float frameTime, SupermatterComponent? SMcomponent = null, TransformComponent? Xform = null)
+        {
+            if (!Resolve(Uid, ref SMcomponent, ref Xform))
+            {
+                return;
+            }
+
+            SMcomponent.AtmosUpdateAccumulator += frameTime;
+            if (_atmosphereSystem.GetTileMixture(Xform.Coordinates) is { } mixture && SMcomponent.AtmosUpdateAccumulator > SupermatterComponent.AtmosUpdateTimer)
```

Comment:
check atmosupdateaccumulator first so it short circuits and doesn't do the costly `GetTileMixture` when it doesn't need to